bigtable/admin/admin_client.cc

```diff
@@ -0,0 +1,89 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/admin/admin_client.h"
+
+#include <absl/memory/memory.h>
+
+namespace {
+/// An implementation of the bigtable::AdminClient interface.
+class SimpleAdminClient : public bigtable::AdminClient {
+ public:
+  SimpleAdminClient(std::string project, bigtable::ClientOptions options)
+      : project_(std::move(project)), options_(std::move(options)) {}
+
+  std::string const& project() const override { return project_; }
+  void on_completion(grpc::Status const& status) override;
+  ::google::bigtable::admin::v2::BigtableTableAdmin::StubInterface&
+  table_admin() override;
+
+ private:
+  void refresh_credentials_and_channel();
+
+ private:
+  std::string project_;
+  bigtable::ClientOptions options_;
+  mutable std::mutex mu_;
```

Comment:
Please add `GUARDED_BY` annotations (via Abseil's [`thread_annotations.h`](https://github.com/abseil/abseil-cpp/blob/master/absl/base/thread_annotations.h)) to ensure that:

* it's clear what member vars are guarded by `mu_` and
* the locks are enforced at compile time

Reply:
Done.  But there are bigger problems here.